### PR TITLE
WIP: Record: Copy Network + Dict Residual (LZ77 + Brotli compression-inspired)

### PR DESCRIPTION
**Pull request by pinnerwt**

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — WIP: Record: Copy Network + Dict Residual (LZ77 + Brotli compression-inspired)

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache


Head SHA: 1c54e9d6ea9cbfca1a3d7c4631ceadc0ad77b8fd
PR: #392 — "WIP: Record: Copy Network + Dict Residual (LZ77 + Brotli compression-inspired)"

---

## Check 1 — N-gram family bug (CLOSE trigger: target token in hash key)

**Result: CLEAN**

The trigram hash key is constructed as:

```python
key = (int(tokens[i-2]) * 1024 + int(tokens[i-1])) % trigram_buckets
trigram[key, tokens[i]] += 1
```

The key uses only the two *context* tokens (`i-2`, `i-1`). The target token `tokens[i]` appears only as the column index in the count accumulation — it is not included in the key. This is the correct causal construction. The banned pattern (including the target token in the hash key) is absent.

The bigram table is a standard `[prev, curr]` matrix; `prev_ids` is derived from `input_ids[:, :-1]` shifted right — no target leakage. The trigram buffer (`log_trigram`) is registered but not used in the forward pass; the forward only indexes `log_bigram[prev_ids]`. No issue.

---

## Check 2 — Pre-Quant TTT (CLOSE trigger: multi-epoch AdamW on val_tokens without score-first)

**Result: CLEAN — No TTT of any kind present**

There is no test-time training, fine-tuning on validation tokens, or any optimizer step that touches `val_tokens`. The validation path (`eval_val`) runs entirely under `torch.inference_mode()` and `model.eval()`, with no gradient computation or optimizer calls. No AdamW or any optimizer references val data at any point in the script.

---

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source. If this review misread your code, please call it out so I can re-audit manually.*

**Comment by MatoTeziTanka** [comment]
# PR 392: Copy Network + Dict Residual (WIP)

## Review Summary

**PR Title:** WIP: Record: Copy Network + Dict Residual (LZ77 + Brotli compression-inspired)
**Status:** OPEN (WIP) | No reviewer comments
**Train File:** records/track_10min_16mb/2026-03-22_CopyNet_DictResidual/train_gpt.py
**Classification:** PURE_NEURAL_CLEAN

## Red Flag Analysis

| Signal | Finding |
|--------|---------|
| target-in-key loss | CLEAN - No custom loss dict indexing |
| TTT/SLOT classes | CLEAN - Not detected |
| Custom tokenizer | CLEAN - Standard SentencePiece |
| loss[] indexing | CLEAN - Standard BPB metric |

## Findings

1. **WIP Status**: Marked as Work In Progress - may still be in development
2. **Approach**: Copy mechanism + learnable dictionary for residual compression (inspired by LZ77/Brotli)
3. **Architecture exploration**: Copy networks are standard neural technique (no loss gaming)
4. **Path**: `records/track_10min_16mb/2026-03-22_CopyNet_DictResidual/`

## Recommendation

**MERGE** - WIP status understood. Copy networks are legitimate architectural approach. No custom tokenizer or loss manipulation detected.


**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** MERGE-pending standard record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Classification via sibling-session agent (Haiku-backed). This review was drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
